main_f4.c

```diff
@@ -411,6 +412,8 @@ main(void)
 	rcc_clock_setup_hse_3v3(&clock_setup);
 
 	/* start the interface */
+	char * sn = "SOME_SERIAL_NO";
```

Comment:
Its fairly simple: Just do a while loop for the string, pull 4 bit out of a serial byte and if between 0 and 9 the character will be '0' + x and if > 9 its 'A' + x - ('9' - '0'). You just have to realise that characters are in the ASCII table and can be used to do offset math. Then fill in the resulting char number into each field, done.